The ProcessAssertion object should hold a AssertionCapability object

pvollan opened the pull request:
#### c0ce523ac7bf189cb5b130204736ac936c4c9312
<pre>
The ProcessAssertion object should hold a AssertionCapability object
<a href="https://bugs.webkit.org/show_bug.cgi?id=268343">https://bugs.webkit.org/show_bug.cgi?id=268343</a>
<a href="https://rdar.apple.com/121614924">rdar://121614924</a>

Reviewed by Chris Dumez.

The ProcessAssertion object should hold a AssertionCapability object, and not a platform _SECapability object.
Otherwise, the invalidation blocks will be deleted in the ProcessAssertion constructor, since we only created
a temporary object there.

* Source/WebKit/Platform/cocoa/AssertionCapability.h:
* Source/WebKit/Platform/cocoa/AssertionCapability.mm:
(WebKit::AssertionCapability::platformCapability const):
* Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
(WebKit::ProcessAssertion::ProcessAssertion):
(WebKit::ProcessAssertion::acquireSync):
* Source/WebKit/UIProcess/ProcessAssertion.h:

Canonical link: <a href="https://commits.webkit.org/273769@main">https://commits.webkit.org/273769@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/46c4894fdf1b2613ecbf7e1466e9b0d9ef733e26

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36514 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15453 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38733 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39178 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32755 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17922 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12535 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31385 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37074 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13024 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32308 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11406 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11429 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40424 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33089 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32900 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37352 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11660 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9516 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35456 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13356 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8301 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12097 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12545 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
```diff
@@ -392,9 +392,9 @@ static ASCIILiteral runningBoardDomainForAssertionType(ProcessAssertionType asse
         };
         AssertionCapability capability { process.environmentIdentifier(), runningBoardDomain, runningBoardAssertionName, WTFMove(willInvalidateBlock), WTFMove(didInvalidateBlock) };
-        m_capability = capability.platformCapability();
+        m_capability = WTFMove(capability);
```
I would have done this:
```
m_capability  = AssertionCapability { process.environmentIdentifier(), runningBoardDomain, runningBoardAssertionName, WTFMove(willInvalidateBlock), WTFMove(didInvalidateBlock) };
```

pvollan replied:
Ah, good point, will fix!

Thanks for reviewing!

cdumez commented on Source/WebKit/Platform/cocoa/AssertionCapability.mm:
```diff
@@ -39,17 +39,17 @@
-    return nil;
+}
+
+RetainPtr<_SECapability> AssertionCapability::platformCapability() const
```
I think this should return ` _SECapability *`

pvollan replied:
Would you be ok with keeping this as-is, since this is a virtual method, and changing this would be a bigger change?

Thanks for reviewing!

cdumez replied:
Ok.

pvollan replied:
Thank you Chris!

cdumez commented on Source/WebKit/Platform/cocoa/AssertionCapability.mm:
```diff
@@ -39,17 +39,17 @@
+
+RetainPtr<_SECapability> AssertionCapability::platformCapability() const
+{
+    return m_capability;
```
`.get()`

cdumez commented on Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
```diff
@@ -466,9 +465,9 @@ static ASCIILiteral runningBoardDomainForAssertionType(ProcessAssertionType asse
+    if (m_process && m_capability) {
         NSError *error = nil;
-        m_grant = [m_process grantCapability:m_capability.get() error:&error];
+        m_grant = [m_process grantCapability:m_capability->platformCapability().get() error:&error];
```
Remove `.get()`

cdumez reviewed (approved):
LGTM with nits.

pvollan committed:
The ProcessAssertion object should hold a AssertionCapability object
https://bugs.webkit.org/show_bug.cgi?id=268343
rdar://121614924

Reviewed by Chris Dumez.

The ProcessAssertion object should hold a AssertionCapability object, and not a platform _SECapability object.
Otherwise, the invalidation blocks will be deleted in the ProcessAssertion constructor, since we only created
a temporary object there.

* Source/WebKit/Platform/cocoa/AssertionCapability.h:
* Source/WebKit/Platform/cocoa/AssertionCapability.mm:
(WebKit::AssertionCapability::platformCapability const):
* Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
(WebKit::ProcessAssertion::ProcessAssertion):
(WebKit::ProcessAssertion::acquireSync):
* Source/WebKit/UIProcess/ProcessAssertion.h:

Canonical link: https://commits.webkit.org/273769@main
